flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTableSourceScan.scala

```diff
@@ -153,33 +155,91 @@ class StreamExecTableSourceScan(
 
     val ingestedTable = new DataStream(planner.getExecEnv, streamTransformation)
 
-    // generate watermarks for rowtime indicator
-    val rowtimeDesc: Option[RowtimeAttributeDescriptor] =
-      TableSourceUtil.getRowtimeAttributeDescriptor(tableSource, tableSourceTable.selectedFields)
-
-    val withWatermarks = if (rowtimeDesc.isDefined) {
-      val rowtimeFieldIdx = getRowType.getFieldNames.indexOf(rowtimeDesc.get.getAttributeName)
-      val watermarkStrategy = rowtimeDesc.get.getWatermarkStrategy
-      watermarkStrategy match {
-        case p: PeriodicWatermarkAssigner =>
-          val watermarkGenerator = new PeriodicWatermarkAssignerWrapper(rowtimeFieldIdx, p)
-          ingestedTable.assignTimestampsAndWatermarks(watermarkGenerator)
-        case p: PunctuatedWatermarkAssigner =>
-          val watermarkGenerator =
-            new PunctuatedWatermarkAssignerWrapper(rowtimeFieldIdx, p, producedDataType)
-          ingestedTable.assignTimestampsAndWatermarks(watermarkGenerator)
-        case _: PreserveWatermarks =>
-          // The watermarks have already been provided by the underlying DataStream.
-          ingestedTable
-      }
+    val withWatermarks = if (tableSourceTable.watermarkSpec.isDefined) {
+      // generate watermarks for rowtime indicator from WatermarkSpec
+      val rowType = tableSourceTable.getRowType(planner.getTypeFactory)
+      val converter = planner.plannerContext.createSqlToRexConverter(rowType)
+      applyWatermarkByWatermarkSpec(
+        config,
+        ingestedTable,
```

Comment:
Can we unify the logic with `RowtimeAttributeDescriptor `, then we can remove the "if else" clause.

Reply:
Unfortunately, we can't. The `WatermarkStrategy` in `RowtimeAttributeDescriptor` is an imperative API, it's hard to convert it into a `RexNode` or UDF.